src/lib/yang/data.lua

```diff
@@ -803,11 +803,19 @@ function data_printer_from_grammar(production)
          end
       end
    end
+   print_default = function ()
+      local status = false
+      return function (val)
+         if not val then return status end
+         status = val
+         return status
+      end
+   end
    function handlers.scalar(keyword, production)
       local serialize = value_serializer(production.argument_type)
       return function(data, file, indent)
          local str = serialize(data)
-         if str ~= production.default then
+         if print_default() or str ~= production.default then
```

Comment:
Here print_default() always returns a true value AFAIU -- it returns the function from line 808.  Anyway this doesn't look right to me: can you change instead to add a "print_default" option on line 707?  The idea being, you derive a printer from a grammar either with default-printing or without.  It's a property of the printer.  That way you can just check a simple variable in the scalar printer to know whether to print defaults or no.